README.md
# NAME

MooseX::Traits - Automatically apply roles at object creation time

# VERSION

version 0.12

# SYNOPSIS

Given some roles:

    package Role;
    use Moose::Role;
    has foo => ( is => 'ro', isa => 'Int' required => 1 );

And a class:

    package Class;
    use Moose;
    with 'MooseX::Traits';

Apply the roles to the class at `new` time:

    my $class = Class->with_traits('Role')->new( foo => 42 );

Then use your customized class:

    $class->isa('Class'); # true
    $class->does('Role'); # true
    $class->foo; # 42

# DESCRIPTION

Often you want to create components that can be added to a class
arbitrarily.  This module makes it easy for the end user to use these
components.  Instead of requiring the user to create a named class
with the desired roles applied, or apply roles to the instance
one-by-one, he can just create a new class from yours with
`with_traits`, and then instantiate that.

There is also `new_with_traits`, which exists for compatibility
reasons.  It accepts a `traits` parameter, creates a new class with
those traits, and then instantiates it.

    Class->new_with_traits( traits => [qw/Foo Bar/], foo => 42, bar => 1 )

returns exactly the same object as

    Class->with_traits(qw/Foo Bar/)->new( foo => 42, bar => 1 )

would.  But you can also store the result of `with_traits`, and call
other methods:

    my $c = Class->with_traits(qw/Foo Bar/);
    $c->new( foo => 42 );
    $c->whatever( foo => 1234 );

And so on.

# METHODS

- __$class->with\_traits( @traits )__

    Return a new class with the traits applied.  Use like:

- __$class->new\_with\_traits(%args, traits => \\@traits)__

    `new_with_traits` can also take a hashref, e.g.:

        my $instance = $class->new_with_traits({ traits => \@traits, foo => 'bar' });

# ATTRIBUTES YOUR CLASS GETS

This role will add the following attributes to the consuming class.

## \_trait\_namespace

You can override the value of this attribute with `default` to
automatically prepend a namespace to the supplied traits.  (This can
be overridden by prefixing the trait name with `+`.)

Example:

    package Another::Trait;
    use Moose::Role;
    has 'bar' => (
        is       => 'ro',
        isa      => 'Str',
        required => 1,
    );

    package Another::Class;
    use Moose;
    with 'MooseX::Traits';
    has '+_trait_namespace' => ( default => 'Another' );

    my $instance = Another::Class->new_with_traits(
        traits => ['Trait'], # "Another::Trait", not "Trait"
        bar    => 'bar',
    );
    $instance->does('Trait')          # false
    $instance->does('Another::Trait') # true

    my $instance2 = Another::Class->new_with_traits(
        traits => ['+Trait'], # "Trait", not "Another::Trait"
    );
    $instance2->does('Trait')          # true
    $instance2->does('Another::Trait') # false

# AUTHOR

Jonathan Rockway <jrockway@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by Infinity Interactive, Inc. http://www.iinteractive.com.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

# CONTRIBUTORS

- Alexander Hartmaier <abraxxa@cpan.org>
- Chris Prather <chris@prather.org>
- Florian Ragwitz <rafl@debian.org>
- Hans Dieter Pearcey <hdp@weftsoar.net>
- Jesse Luehrs <doy@tozt.net>
- Karen Etheridge <ether@cpan.org>
- Matt S. Trout <mst@shadowcatsystems.co.uk>
- Rafael Kitover <rkitover@cpan.org>
- Shawn Moore <sartak@bestpractical.com>
- Stevan Little <stevan.little@iinteractive.com>
- Tomas Doran <bobtfish@bobtfish.net>
- Yuval Kogman <nothingmuch@woobling.org>


Changes
Revision history for MooseX-Traits

{{$NEXT}}

0.13      2015-08-16 04:10:59Z
        - update some distribution tooling

0.12      2013-11-13 07:21:32Z
        - removed use of deprecated Class::MOP::load_class
        - repository migrated from shadowcat to the github moose organization
        - uses of Test::Exception converted to Test::Fatal

0.11   2010-05-13 12:31:00Z
        - Fix doc typo. jawnsy++

0.10   2010-05-12 15:53:56Z
        - Stop pretending the created classes are anonymous.
          They used to close over their own metaclass, so they were never
          cleaned up anyway and only made debugging unnecessarily hard as
          all instances referenced the metaclass as well.
        - Require newer version of MX::Role::Parameterized in the parameterized
          role test. (RT#49873).

0.09   2010-04-07 03:51:40Z
        - Class->with_traits(...)->new is now preferred over
          Class->new_with_traits(...).

0.08   2010-02-13 17:05:23Z
        - Allow you to pass traits => 'OneTrait' rather than having to say
          traits => ['OneTrait'] to help when getting traits from things that
          want to send a scalar (e.g. Config::General)

0.07   2009-08-16 15:38:47Z
        - deprecate apply_traits ("no warnings" to disable warning)
        - add MooseX::Traits::Util so that other modules can resolve trait names
        - add new_class_with_traits util function to create classes with traits

0.06   2009-06-28 22:30:58Z
        - use "is => bare" on newer moosen (Closes: RT#47011) (hdp)
        - depend on a Moose version that provides
          meta->constructor_name (Closes RT#46936)

0.05   2009-05-28 15:37:21Z
        - support subclasses of classes that use MX::Traits

0.04   2009-05-27 06:00:29Z
        - add apply_traits method
        - use meta->constructor_name instead of 'new'

0.03   2008-06-12 16:57:29Z
        - stevan changed some stuff

0.02   2008-06-02 23:40:55Z
        - leave "new" alone, add a new "new_with_traits" instead

0.01   2008-05-24 23:54:26Z
        - inital release


lib/MooseX/Traits/Util.pm
package MooseX::Traits::Util;
use strict;
use warnings;

our $VERSION = '0.14';

use Sub::Exporter -setup => {
    exports => ['new_class_with_traits'],
};

use Class::Load ();
use Carp ();

# note: "$class" throughout is "class name" or "instance of class
# name"

sub check_class {
    my $class = shift;

    Carp::confess "We can't interact with traits for a class ($class) ".
      "that does not do MooseX::Traits" unless $class->does('MooseX::Traits');
}

sub transform_trait {
    my ($class, $name) = @_;
    return $1 if $name =~ /^[+](.+)$/;

    check_class($class);

    my $namespace = $class->meta->find_attribute_by_name('_trait_namespace');
    my $base;
    if($namespace->has_default){
        $base = $namespace->default;
        if(ref $base eq 'CODE'){
            $base = $base->();
        }
    }

    return $name unless $base;
    return join '::', $base, $name;
}

sub resolve_traits {
    my ($class, @traits) = @_;

    check_class($class);

    return map {
        my $orig = $_;
        if(!ref $orig){
            my $transformed = transform_trait($class, $orig);
            Class::Load::load_class($transformed);
            $transformed;
        }
        else {
            $orig;
        }
    } @traits;
}

my $anon_serial = 0;

sub new_class_with_traits {
    my ($class, @traits) = @_;

    check_class($class);

    my $meta;
    @traits = resolve_traits($class, @traits);
    if (@traits) {
        $meta = $class->meta->create(
            join(q{::} => 'MooseX::Traits::__ANON__::SERIAL', ++$anon_serial),
            superclasses => [ $class->meta->name ],
            roles        => \@traits,
            cache        => 1,
        );
    }

    # if no traits were given just return the class meta
    return $meta ? $meta : $class->meta;
}

1;

=for Pod::Coverage check_class new_class_with_traits resolve_traits transform_trait

=cut


lib/MooseX/Traits.pm
package MooseX::Traits;
# ABSTRACT: Automatically apply roles at object creation time

our $VERSION = '0.14';

use Moose::Role;

use MooseX::Traits::Util qw(new_class_with_traits);

use warnings;
use warnings::register;

use namespace::autoclean;

has '_trait_namespace' => (
    # no accessors or init_arg
    init_arg => undef,
    isa      => 'Str',
    is       => 'bare',
);

sub with_traits {
    my ($class, @traits) = @_;

    my $new_class = new_class_with_traits(
        $class,
        @traits,
    );

    return $new_class->name;
}

# somewhat deprecated, but use if you want to
sub new_with_traits {
    my $class = shift;

    my ($hashref, %args) = 0;
    if (ref($_[0]) eq 'HASH') {
        %args    = %{ +shift };
        $hashref = 1;
    } else {
        %args    = @_;
    }

    my $traits = delete $args{traits} || [];

    my $new_class = $class->with_traits(ref $traits ? @$traits : $traits );

    my $constructor = $new_class->meta->constructor_name;
    confess "$class ($new_class) does not have a constructor defined via the MOP?"
      if !$constructor;

    return $new_class->$constructor($hashref ? \%args : %args);

}

# this code is broken and should never have been added.  i probably
# won't delete it, but it is definitely not up-to-date with respect to
# other features, and never will be.
#
# runtime role application is fundamentally broken.  if you really
# need it, write it yourself, but consider applying the roles before
# you create an instance.

=for Pod::Coverage apply_traits

=cut

sub apply_traits {
    my ($self, $traits, $rebless_params) = @_;

    # disable this warning with "use MooseX::Traits; no warnings 'MooseX::Traits'"
    warnings::warnif('apply_traits is deprecated due to being fundamentally broken. '.
                     q{disable this warning with "no warnings 'MooseX::Traits'"});

    # arrayify
    my @traits = $traits;
    @traits = @$traits if ref $traits;

    if (@traits) {
        @traits = MooseX::Traits::Util::resolve_traits(
            $self, @traits,
        );

        for my $trait (@traits){
            $trait->meta->apply($self, rebless_params => $rebless_params || {});
        }
    }
}

no Moose::Role;

1;

__END__

=pod

=head1 SYNOPSIS

Given some roles:

  package Role;
  use Moose::Role;
  has foo => ( is => 'ro', isa => 'Int' required => 1 );

And a class:

  package Class;
  use Moose;
  with 'MooseX::Traits';

Apply the roles to the class at C<new> time:

  my $class = Class->with_traits('Role')->new( foo => 42 );

Then use your customized class:

  $class->isa('Class'); # true
  $class->does('Role'); # true
  $class->foo; # 42

=head1 DESCRIPTION

Often you want to create components that can be added to a class
arbitrarily.  This module makes it easy for the end user to use these
components.  Instead of requiring the user to create a named class
with the desired roles applied, or apply roles to the instance
one-by-one, he can just create a new class from yours with
C<with_traits>, and then instantiate that.

There is also C<new_with_traits>, which exists for compatibility
reasons.  It accepts a C<traits> parameter, creates a new class with
those traits, and then instantiates it.

   Class->new_with_traits( traits => [qw/Foo Bar/], foo => 42, bar => 1 )

returns exactly the same object as

   Class->with_traits(qw/Foo Bar/)->new( foo => 42, bar => 1 )

would.  But you can also store the result of C<with_traits>, and call
other methods:

   my $c = Class->with_traits(qw/Foo Bar/);
   $c->new( foo => 42 );
   $c->whatever( foo => 1234 );

And so on.

=head1 METHODS

=over 4

=item B<< $class->with_traits( @traits ) >>

Return a new class with the traits applied.  Use like:

=item B<< $class->new_with_traits(%args, traits => \@traits) >>

C<new_with_traits> can also take a hashref, e.g.:

  my $instance = $class->new_with_traits({ traits => \@traits, foo => 'bar' });

=back

=head1 ATTRIBUTES YOUR CLASS GETS

This role will add the following attributes to the consuming class.

=head2 _trait_namespace

You can override the value of this attribute with C<default> to
automatically prepend a namespace to the supplied traits.  (This can
be overridden by prefixing the trait name with C<+>.)

Example:

  package Another::Trait;
  use Moose::Role;
  has 'bar' => (
      is       => 'ro',
      isa      => 'Str',
      required => 1,
  );

  package Another::Class;
  use Moose;
  with 'MooseX::Traits';
  has '+_trait_namespace' => ( default => 'Another' );

  my $instance = Another::Class->new_with_traits(
      traits => ['Trait'], # "Another::Trait", not "Trait"
      bar    => 'bar',
  );
  $instance->does('Trait')          # false
  $instance->does('Another::Trait') # true

  my $instance2 = Another::Class->new_with_traits(
      traits => ['+Trait'], # "Trait", not "Another::Trait"
  );
  $instance2->does('Trait')          # true
  $instance2->does('Another::Trait') # false

=cut
